src/main/java/com/google/sps/servlets/VoteServlet.java

```diff
@@ -0,0 +1,113 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.meltingpot.servlets;
+
+import com.google.firebase.auth.FirebaseToken;
+import com.google.gson.Gson;
+import com.google.sps.meltingpot.auth.Auth;
+import com.google.sps.meltingpot.data.DBInterface;
+import com.google.sps.meltingpot.data.DBUtils;
+import com.google.sps.meltingpot.data.FirestoreDB;
+import com.google.sps.meltingpot.data.RecipeMetadata;
+import com.google.sps.meltingpot.data.User;
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Handles recipe vote requests. */
+@WebServlet("/api/vote")
+public class VoteServlet extends HttpServlet {
+  private Gson gson = new Gson();
+  private DBInterface db;
+
+  public VoteServlet(DBInterface db) {
+    this.db = db;
+  }
+
+  public VoteServlet() {}
+
+  @Override
+  public void init() {
+    db = new FirestoreDB();
+  }
+
+  @Override
+  public void doPut(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String recipeId = request.getParameter("recipeId");
+    String vote = request.getParameter("vote");
+    String token = request.getParameter("token");
+
+    if (recipeId == null || vote == null || (!vote.equals("true") && !vote.equals("false"))) {
+      response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+      return;
+    }
+
+    if (!db.isDocument(recipeId, DBUtils.DB_RECIPES_COLLECTION)) {
+      // recipe does not exist
+      response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+      return;
+    }
+
+    FirebaseToken authToken = Auth.verifyIdToken(token);
+    if (authToken == null) {
```

Comment:
Not sure if the code exists on this branch, but this process can be combined with line 79. If Auth.getUid(String token) returns null, then SC_UNAUTHORIZED will already be set and you can just return. Otherwise, you have your uid.

Reply:
Doesn't exist yet, will pull once that gets merged into master.